### DISCUSSION TITLE
Extending existing rules

### DISCUSSION THREAD

**Discussion by jsmolinski-vivint** [detekt/detekt#7072] [Q&A] [answered]
Hi all,

I'm working on updating our Detekt configuration to be more friendly for Compose. One issue that our developers are running into is the LongParameterList rule. I've reviewed the guidance and suggestions [here](https://detekt.dev/docs/introduction/compose/#longparameterlist-for-compose), but was wondering if there was a way we could actually extend the `LongParameterList` class (and potentially others) and provide an alternate check (or omit the check) for Compose-annotated functions. 

I tried creating a custom rule and extending the `LongParameterList` rule, but it's closed/final. I'm able to hack around it by creating an "outer-rule" and selectively passing findings from `LongParameterList` as an "inner-rule", but it's pretty convoluted. 

So I have two questions:
1) Would it be possible to open the built-in rules for developers to extend and customize their behavior? 
2) If not, is there another way to extend existing rules without re-implementing them or creating this strange "outer-rule" and "inner-rule" relationship?

Thank you!

**Comment by schalkms** [comment, marked as answer]
I don't understand what you mean by outer- and inner-rule.

Nevertheless, the suggested approach is to copy the source code of `LongParameterList` as it is to a custom rule and modify the needed parts.

Just setting classes of built-in rules to open doesn't provide a remedy, because you also need to provide hooks for all possible parts needing to be customized. I don't think it's worth the effort.

**Reply by jsmolinski-vivint** [reply]
Sure, that makes sense. Thanks for your reply!

**Comment by cortinico** [comment]
> Nevertheless, the suggested approach is to copy the source code of LongParameterList as it is to a custom rule and modify the needed parts.

That's the suggested approach.

If there is a snippet of code that is frequently re-used to manipulate the PSI, we can consider exposing it as a public API (i.e. see the `detekt-utils` artifact). 

But we don't offer an easy way to 'extend' rules as you're suggesting as that will make our API surface explode
